iiasa/ixmp
# Transfer Excel data I/O from message_ix

This PR streamlines and improves file input/output in the MESSAGE stack:

- `Scenario.read_excel()`, `.to_excel()` are brought to ixmp from message_ix (see iiasa/message_ix#321).
  - Can now be used with any ixmp Scenario.
  - The code is improved in several ways, e.g. it can now distinguish between 1-D index sets and indexed sets; degrades gracefully by giving readable log messages and exceptions.
- `TimeSeries.read_file` is added as a replacement for `ixmp.utils.import_timeseries`.
- New: commands `import` and `export` that allow performing the above operations from the command-line.
- New: documentation that clearly describes the file formats supported (and not supported) for I/O; docstrings for all touched methods.
- New: `TimeSeries.add_timeseries(..., year_lim=(2020, 2080))`, which brings the minimum/maximum year functionality previously only supported by `import_timeseries()` to non-file-based operations.
- See #290 & thanks @zikolach for:
  - ixmp.jar is updated to clean up Java exception handling.
  - {,JDBC}Backend gain a method, `get_log_level()`.

Closes #284.
Replaces #286.

## How to review

- Build and read the documentation.
- Check that any I/O scripts still work.

  One possible breaking change: `import_timeseries()` previously automatically prepended the string 'R11_' to imported region names, unless the name was 'World'. This is specific to MESSAGEix-GLOBIOM (not message_ix), so it is removed.

## PR checklist
